src/terms.jl

```diff
@@ -0,0 +1,572 @@
+abstract type AbstractTerm end
+const TupleTerm = NTuple{N, AbstractTerm} where N
+const TermOrTerms = Union{AbstractTerm, NTuple{N, AbstractTerm} where N}
+
+function Base.show(io::IO, mime::MIME"text/plain", term::AbstractTerm)
+    print(io, get(io, :prefix, ""), term)
+end
+
+function Base.show(io::IO, mime::MIME"text/plain", terms::TupleTerm)
+    for t in terms
+        show(io, mime, t)
+        # ensure that there are newlines in between each term after the first
+        # if no prefix is specified
+        io = IOContext(io, :prefix => get(io, :prefix, "\n"))
```

Comment:
Is this `prefix` system really needed? I'm not sure using non-standard `IOContext` properties like this is recommended, since in theory it could affect the printing of any object inside terms (e.g. the symbols or numbers).

Reply:
I don't think it's NEEDED, and actually after going through the bother of implementing it this way it doesn't clean up the printing logic as much as I'd hoped.  (I'd hoped it would make it possible to avoid lots of loops and special logic and just use the same kind of recursive `show`ing of child terms but it didn't really work out that way...)

Reply:
I've looked at this a little more, and taking the prefix out of the IOContext would require providing it some other way, like an internal `_show_long` method that has an argument for the prefix (this is like what AbstractTrees.jl does).  But that seems annoying and kind of hostile to extensibility.  The current solution for extending (a fallback for `show(::AbstractTerm, ::MIME"text/plain", ::IO)` that just calls `print` with the prefix) isn't _awesome_ though, and there's nothing preventing doing something similar with `_show_long` or whatever...

Reply:
What if you pass the keyword argument directly to `show`? Custom term types would just have to accept it.

Reply:
Ah yes that could work too.  I was so excited to use `IOContext` finally though ;)